gdrapi.c

```diff
@@ -392,15 +498,24 @@ int gdr_copy_from_bar(void *h_ptr, const void *bar_ptr, size_t size)
             break;
         }
 
-        // fall through
-        gdr_dbgc(1, "using plain implementation of gdr_copy_from_bar\n");
-        memcpy(h_ptr, bar_ptr, size);
-
+        // on POWER, compiler memcpy is not optimal for MMIO
+        // 64bit loads have 2x the BW of 32bit ones
+        if (PREFERS_LOAD_UNROLL8 && is_aligned(size, 8) && ptr_is_aligned(bar_ptr, 8) && ptr_is_aligned(h_ptr, 8)) {
+            gdr_dbgc(1, "using unroll8_memcpy for gdr_copy_from_bar\n");
+            unroll8_memcpy(h_ptr, bar_ptr, size);
+            break;
+        } else if (PREFERS_LOAD_UNROLL4 && is_aligned(size, 4) && ptr_is_aligned(bar_ptr, 4) && ptr_is_aligned(h_ptr, 4)) {
+            gdr_dbgc(1, "using unroll4_memcpy for gdr_copy_from_bar\n");
+            unroll4_memcpy(h_ptr, bar_ptr, size);
+            break;
+        } else {
+            gdr_dbgc(1, "using plain memcpy for gdr_copy_from_bar\n");
+            memcpy(h_ptr, bar_ptr, size);
+        }
         // note: fencing is not needed because plain stores are used
```

Comment:
Isn't this a load operation -- i.e., load from bar?

Reply:
right, but if the store to user-supplied buffer was non-temporal, e.g. on x86, then a fence would be needed.